### PR TITLE
Update Yarn to v4

### PR DESCRIPTION
**Pull request by noahtallen**
This PR updates Yarn to the latest stable version. I did this following [these instructions](https://yarnpkg.com/getting-started/install):

```sh
yarn set version stable
yarn install
```

See here: https://yarnpkg.com/advanced/changelog. We're updating from 3.2 to 4.

The vast majority of changes won't impact us, but there are some improvements to Yarn workspaces that might be useful. (I ran into some issues working on the combined calypso app builds).

Testing:
- CI
- Run `yarn install` and test local development workflows.

### CODE REVIEW

File: .yarnrc.yml [hunk 1]
```diff
@@ -2,8 +2,7 @@
 compressionLevel: 0
 nodeLinker: node-modules
 enableGlobalCache: true
-preferAggregateCacheInfo: true
```
Comment (noahtallen) [inline .yarnrc.yml:5]: This is now always enabled

**Review by tyxla — COMMENTED** [review]
Are we still interested in any of the improvements you were referring to?

Regardless, this looks good to me, but we might need a rebase before shipping.

**Comment by noahtallen** [comment]
So the workspaces documentation here (https://yarnpkg.com/cli/workspaces/foreach) mentions a worktree option: https://yarnpkg.com/cli/workspaces/foreach#options-W%2C-worktree I was hoping we could use this to make Yarn automatically find the apps we want to build. However, it still says it's an unsupported option even after updating Yarn. 

I used something else, so I don't need this option any more. But still a good idea to update :)

File: .yarnrc.yml [hunk 2]
```diff
@@ -2,12 +2,7 @@
...
-yarnPath: .yarn/releases/yarn-3.2.3.cjs
-
-plugins:
-  - path: .yarn/plugins/@yarnpkg/plugin-workspace-tools.cjs
-    spec: "@yarnpkg/plugin-workspace-tools"
```
Comment (noahtallen) [inline .yarnrc.yml:10]: These plugins are included by default

**Comment by noahtallen** [comment]
I think we have two issues which are related:
1. We need to fix the warnings by updating `.yarnrc.yml`
2. We need to investigate why `yarn install --json` doesn't print anything. This is used to fail CI when there are warnings as in step 1, so our protection against introducing further warnings is broken.

opened a bug report: https://github.com/yarnpkg/berry/issues/5866

**Comment by noahtallen** [comment]
I updated some of the log filters, and was surprisingly able to remove several of them as well! `yarn install` should show no warnings/errors now. 

However, `yarn install --json` is still broken.

**Comment by noahtallen** [comment]
I tried to use corepack for Yarn, removing the `installPath` and `engines` options we had set. But it seems to be flaky -- it worked for me today, but not for others. Since it's experimental, I suppose we should keep installPath for now

**Comment by arcanis** [comment]
> it worked for me today, but not for others

Do you have details on the errors other people got? It's experimental in the sense that we may want to tweak the CLI before marking it stable, but apart from that it's supposed to be "production-ready" in terms of stability.

**Comment by flootr** [comment]
Here's the error message we got:

```
node:internal/modules/cjs/loader:1080
  throw err;
  ^

Error: Cannot find module '/wp-calypso/.yarn/releases/yarn-3.1.1.cjs'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
    at Module._load (node:internal/modules/cjs/loader:922:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v18.17.0
```

**Comment by noahtallen** [comment]
The error happens immediately on machine that hasn't tried this yet, with commit https://github.com/Automattic/wp-calypso/pull/82688/commits/680548be92c8126deb6ea40af63414ac98ea4d96. 

- Checkout that commit
- `yarn install`
- See the error above, even after `corepack enable`, and even in a new shell
- Check out the head of the branch (which reverts that change), and `yarn install` works
- Check out the bad commit again in a new shell and run `corepack enable`
- Then `yarn install` does work

It almost seems like a caching issue where corepack (or whatever mechanism) is still trying to use yarn 3, despite everything in this branch pointing to yarn 4

**Comment by noahtallen** [comment]
Nice, looks like the JSON flag has been fixed! https://github.com/yarnpkg/berry/pull/5933

We may have to wait a bit more for a release though :)

**Comment by noahtallen** [comment]
Updated to 4.0.2, and `--json` is now working. Our parser for yarn errors is also working in the unit tests: https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/11247344?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildProblemsSection=true&expandBuildChangesSection=true

File: bin/build-packages-if-needed.sh [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+# yarn start fails when this package is not built. It a good indication that the
+# prerequisite packages were cleaned and need to be prepared again.
```
Comment (flootr) [inline bin/build-packages-if-needed.sh:4]: 👍🏻

**Review by flootr — APPROVED** [review]
works great, thank you 👍🏻 🚀

**Pull request merged**